libnetwork/drivers/bridge/setup_ip_tables_linux.go

```diff
@@ -502,3 +511,67 @@ func clearConntrackEntries(nlh *netlink.Handle, ep *bridgeEndpoint) {
+//
+// Without mirrored mode networking, or for a packet sent from Linux, packets
+// sent to 127.0.0.1 are processed as outgoing - they hit the nat-OUTPUT chain,
+// which does not jump to the nat-DOCKER chain because the rule has an exception
+// for "-d 127.0.0.0/8". The default action on the nat-OUTPUT chain is ACCEPT (by
+// default), so the packet is delivered to 127.0.0.1 on lo, where docker-proxy
+// picks it up and acts as a man-in-the-middle; it receives the packet and
+// re-sends it to the container (or acks a SYN and sets up a second TCP
+// connection to the container). So, the container sees packets arrive with a
+// source address belonging to the network's bridge, and it is able to reply to
+// that address.
+//
+// In WSL2's mirrored networking mode, Linux has a loopback0 device as well as lo
+// (which owns 127.0.0.1 as normal). Packets sent to 127.0.0.1 from Windows to a
+// server listening on Linux's 127.0.0.1 are delivered via loopback0, and
+// processed as packets arriving from outside the Linux host (which they are).
+//
+// So, these packets hit the nat-PREROUTING chain instead of nat-OUTPUT. It would
+// normally be impossible for a packet ->127.0.0.1 to arrive from outside the
+// host, so the nat-PREROUTING jump to nat-DOCKER has no exception for it. The
+// packet is processed by a per-bridge DNAT rule in that chain, so it is
+// delivered directly to the container (not via docker-proxy) with source address
+// 127.0.0.1, so the container can't respond.
+//
+// DNAT is normally skipped by RETURN rules in the nat-DOCKER chain for packets
+// arriving from any other bridge network. Similarly, this function adds (or
+// removes) a rule to RETURN early for packets delivered via loopback0 with
+// destination 127.0.0.0/8.
+func mirroredWSL2Workaround(ipv iptables.IPVersion) error {
+	// WSL2 does not (currently) support Windows<->Linux communication via ::1.
+	if ipv != iptables.IPv4 {
+		return nil
+	}
+	return programChainRule(mirroredWSL2Rule(), "WSL2 loopback", mirroredWSL2())
+}
+
+// mirroredWSL2 returns true if the host Linux appears to be running under
+// Windows WSL2 with mirrored mode networking. If a loopback0 device exists, it
+// checks that /usr/bin/wslinfo is executable and reports mirrored networking.
+func mirroredWSL2() bool {
+	if _, err := netlink.LinkByName("loopback0"); err != nil {
+		if !errors.As(err, &netlink.LinkNotFoundError{}) {
+			log.G(context.TODO()).Warnf("Failed to check for WSL interface: %v", err)
+		}
+		return false
+	}
+	output, err := exec.Command(wslinfoPath, "--networking-mode", "-n").CombinedOutput()
+	log.G(context.TODO()).Debugf("wslinfo --networking-mode:%s err:%v", string(output), err)
+	return err == nil && string(output) == "mirrored"
+}
+
+func mirroredWSL2Rule() iptRule {
+	return iptRule{
+		ipv:   iptables.IPv4,
+		table: iptables.Nat,
+		chain: DockerChain,
+		args:  []string{"-i", "loopback0", "-d", "127.0.0.0/8", "-j", "RETURN"},
```

Comment:
As (even more with my suggestions), the magic `loopback0` interface will be used more now, I'm wondering it would make sense to define a `const` for it as well; at least it would allow documenting "what's this magic `loopback0`? 


(downside is that it's less grep'able for `loopback0` in the code, although the `const` can still be a good starting point I guess)

So, yeah, not a "strong" opinion one way of another 😅

Reply:
I tried it with a const name ... but I don't think it made things any clearer, so put it back.

The gigantic comment just above explains what it is.